pkg/apps/server.go

```diff
@@ -47,26 +49,55 @@ func NewSwiftFileServer(conn *swift.Connection, appsType AppType) FileServer {
 
 func (s *swiftServer) Open(slug, version, file string) (io.ReadCloser, error) {
 	objName := s.makeObjectName(slug, version, file)
-	f, _, err := s.c.ObjectOpen(s.container, objName, false, nil)
+	f, h, err := s.c.ObjectOpen(s.container, objName, false, nil)
 	if err != nil {
 		return nil, wrapSwiftErr(err)
 	}
+	o := h.ObjectMetadata()
+	if contentEncoding := o["content-encoding"]; contentEncoding == "gzip" {
+		return gzip.NewReader(f)
+	}
 	return f, nil
 }
 
 func (s *swiftServer) ServeFileContent(w http.ResponseWriter, req *http.Request, slug, version, file string) error {
 	objName := s.makeObjectName(slug, version, file)
-	f, o, err := s.c.ObjectOpen(s.container, objName, false, nil)
+	f, h, err := s.c.ObjectOpen(s.container, objName, false, nil)
 	if err != nil {
 		return wrapSwiftErr(err)
 	}
 	defer f.Close()
 
-	if req.Header.Get("Cache-Control") == "" && req.Header.Get("Range") == "" {
-		w.Header().Set("Etag", fmt.Sprintf(`"%s"`, o["Etag"][:10]))
+	if checkETag := req.Header.Get("Cache-Control") != ""; checkETag {
+		etag := fmt.Sprintf(`"%s"`, h["Etag"][:10])
+		if web_utils.CheckPreconditions(w, req, etag) {
+			return nil
+		}
+		w.Header().Set("Etag", etag)
+	}
+
+	var r io.Reader = f
+	contentLength := h["Content-Length"]
+	contentType := h["Content-Type"]
+	o := h.ObjectMetadata()
+	if contentEncoding := o["content-encoding"]; contentEncoding == "gzip" {
+		if strings.Contains(req.Header.Get("Accept-Encoding"), "gzip") {
+			w.Header().Set("Content-Encoding", "gzip")
+		} else {
+			contentLength = o["original-content-length"]
+			r, err = gzip.NewReader(f)
+			if err != nil {
+				return err
+			}
+		}
 	}
 
-	http.ServeContent(w, req, objName, unixZeroEpoch, f)
+	if contentType == "" {
+		contentType = magic.MIMETypeByExtension(path.Ext(file))
```

Comment:
It looks like we are already trying to guess the content-type before storing the file. Why should we try again when serving a file? Is it for old files (retro-compatibility)?

Reply:
Yes, this is only for retro-compatibility reasons